wgpu-core/src/lock/vanilla.rs

```diff
@@ -56,45 +119,102 @@ impl<T: fmt::Debug> fmt::Debug for Mutex<T> {
     }
 }
 
-/// A plain wrapper around [`parking_lot::RwLock`].
+/// A plain wrapper around [`implementation::RwLock`].
 ///
-/// This is just like [`parking_lot::RwLock`], except that our [`new`]
+/// This is just like [`implementation::RwLock`], except that our [`new`]
 /// method takes a rank, indicating where the new mutex should sit in
 /// `wgpu-core`'s lock ordering. The rank is ignored.
 ///
 /// See the [`lock`] module documentation for other wrappers.
 ///
 /// [`new`]: RwLock::new
 /// [`lock`]: crate::lock
-pub struct RwLock<T>(parking_lot::RwLock<T>);
+pub struct RwLock<T>(implementation::RwLock<T>);
 
 /// A read guard produced by locking [`RwLock`] as a reader.
 ///
-/// This is just a wrapper around a [`parking_lot::RwLockReadGuard`].
-pub struct RwLockReadGuard<'a, T>(parking_lot::RwLockReadGuard<'a, T>);
+/// This is just a wrapper around a [`implementation::RwLockReadGuard`].
+pub struct RwLockReadGuard<'a, T> {
+    guard: implementation::RwLockReadGuard<'a, T>,
+}
 
 /// A write guard produced by locking [`RwLock`] as a writer.
 ///
-/// This is just a wrapper around a [`parking_lot::RwLockWriteGuard`].
-pub struct RwLockWriteGuard<'a, T>(parking_lot::RwLockWriteGuard<'a, T>);
+/// This is just a wrapper around a [`implementation::RwLockWriteGuard`].
+pub struct RwLockWriteGuard<'a, T> {
+    guard: implementation::RwLockWriteGuard<'a, T>,
+    /// Allows for a safe `downgrade` method without `parking_lot`
+    #[cfg(not(feature = "parking_lot"))]
+    lock: &'a RwLock<T>,
+}
 
 impl<T> RwLock<T> {
     pub fn new(_rank: super::rank::LockRank, value: T) -> RwLock<T> {
-        RwLock(parking_lot::RwLock::new(value))
+        RwLock(implementation::RwLock::new(value))
     }
 
     pub fn read(&self) -> RwLockReadGuard<T> {
-        RwLockReadGuard(self.0.read())
+        cfg_if::cfg_if! {
+            if #[cfg(feature = "parking_lot")] {
+                let guard = self.0.read();
+            } else if #[cfg(feature = "std")] {
+                let guard = self.0.read().unwrap_or_else(std::sync::PoisonError::into_inner);
+            } else if #[cfg(feature = "spin")] {
+                let guard = self.0.read();
+            } else {
+                let guard = loop {
+                    if let Ok(guard) = self.0.try_borrow() {
+                        break guard;
+                    }
+                    core::hint::spin_loop();
```

Comment:
This will definitionally never yield as there isn't another thread that can access the RefCell

Reply:
I think i'd prefer if not having either `spin` or `parking_lot` was a compile error. Similarly, do we need to have separate std and parking_lot features?